fdbserver/BlobManifest.actor.cpp

```diff
@@ -90,24 +109,233 @@ struct BlobManifestFile {
+		throw blob_restore_missing_manifest();
+	}
+
+	// Delete all files of oldest manifest
+	static void deleteOldest(const std::vector<BlobManifestFile>& allFiles,
+	                         Reference<BackupContainerFileSystem> container) {
+		if (allFiles.empty()) {
+			return;
+		}
+		int64_t epoch = allFiles.back().epoch;
+		int64_t seqNo = allFiles.back().seqNo;
+		for (auto& f : allFiles) {
+			if (f.epoch == epoch && f.seqNo == seqNo) {
+				container->deleteFile(f.fileName);
+			}
+		}
+	}
+
+	// Count how many manifests
+	static int count(const std::vector<BlobManifestFile>& allFiles) {
+		if (allFiles.empty())
+			return 0;
+
+		int64_t epoch = allFiles.front().epoch;
+		int64_t seqNo = allFiles.front().seqNo;
+		int count = 1;
+		for (auto& f : allFiles) {
+			if (f.epoch != epoch || f.seqNo != seqNo) {
+				count++;
+				epoch = f.epoch;
+				seqNo = f.seqNo;
+			}
+		}
+		return count;
+	}
+
+private:
+	std::vector<BlobManifestFile> files_;
+};
+
+// Splitter could write manifest content into a collection of files with bounded # rows
+class BlobManifestFileSplitter : public ReferenceCounted<BlobManifestFileSplitter> {
+public:
+	BlobManifestFileSplitter(Reference<BlobConnectionProvider> blobConn, int64_t epoch, int64_t seqNo)
+	  : segmentNo_(1), blobConn_(blobConn), epoch_(epoch), seqNo_(seqNo), closed_(false), totalRows_(0),
+	    logicalSize_(0), totalBytes_(0) {}
+
+	// Append a new row to the splitter
+	void append(KeyValueRef row) {
+		ASSERT(!closed_);
+		rows_.push_back_deep(rows_.arena(), row);
+		++totalRows_;
+		logicalSize_ += row.expectedSize();
+		if (logicalSize_ > SERVER_KNOBS->BLOB_RESTORE_MANIFEST_FILE_MAX_SIZE) {
+			flushNext();
+		}
+	}
+
+	// Close the splitter. No more data should be added after
+	ACTOR static Future<Void> close(Reference<BlobManifestFileSplitter> self) {
+		self->flushNext();
+		wait(waitForAll(self->pendingFutures_));
+		self->pendingFutures_.clear();
+		self->flushTailer();
+		wait(waitForAll(self->pendingFutures_));
+		self->closed_ = true;
+		return Void();
+	}
+
+	// Reset writer and start to write from the beginning
+	ACTOR static Future<Void> reset(Reference<BlobManifestFileSplitter> self) {
+		dprint("Reset manifest file {} {}\n", self->epoch_, self->seqNo_);
+		TraceEvent("ResetBlobManifestFile").detail("SegNo", self->segmentNo_);
+		self->rows_.clear();
+		self->segmentNo_ = 1;
+		self->totalRows_ = 0;
+		self->logicalSize_ = 0;
+		wait(waitForAll(self->pendingFutures_));
+		return Void();
+	}
+
+private:
+	// Write next segment
+	void flushNext() {
+		std::string fname = fileName(segmentNo_);
+		Optional<CompressionFilter> compressionFilter;
+		Optional<BlobGranuleCipherKeysCtx> cipherKeysCtx;
+		Value bytes = serializeChunkedSnapshot(StringRef(fname),
+		                                       rows_,
+		                                       SERVER_KNOBS->BG_SNAPSHOT_FILE_TARGET_CHUNK_BYTES,
```

Comment:
We might want a different chunk size parameter for these files because the file size could be very different?

Reply:
I was thinking the chunk size should be same? eventually the chunk size will depend on how small compression/encryption task we want to have.

Reply:
Yeah that's fair actually, makes sense.